README
                                 PLASMA README

      Univ. of Tennessee, Univ. of California Berkeley and Univ. of Colorado
      Denver
   __________ ____       _____    _________   _____      _____
   \______   \    |     /  _  \  /   _____/  /     \    /  _  \
    |     ___/    |    /  /_\  \ \_____  \  /  \ /  \  /  /_\  \
    |    |   |    |___/    |    \/        \/    Y    \/    |    \
    |____|   |_______ \____|__  /_______  /\____|__  /\____|__  /
                     \/       \/        \/         \/         \/

      Parallel Linear Algebra Software for Multicore Architectures

      [1]http://icl.cs.utk.edu/plasma/
     __________________________________________________________________

1. Purpose of PLASMA

   The  main  purpose of PLASMA is to address the performance shortcomings
   of the [2]LAPACK and [3]ScaLAPACK libraries on multicore processors and
   multi-socket  systems  of  multicore  processors and their inability to
   efficiently  utilize  accelerators  such  as  Graphics Processing Units
   (GPUs).  PLASMA  provides  routines  to  solve dense general systems of
   linear   equations,  symmetric  positive  definite  systems  of  linear
   equations and linear least squares problems, using LU, Cholesky, QR and
   LQ factorizations. Real arithmetic and complex arithmetic are supported
   in both single precision and double precision.

   PLASMA has been designed to supercede LAPACK and ScaLAPACK, principally
   by restructuring the software to achieve much greater efficiency, where
   possible,  on  modern  computers  based on multicore processors. PLASMA
   also  relies  on new or improved algorithms. Currently, however, PLASMA
   does  not  serve  as  a  complete  replacement of LAPACK due to limited
   functionality.  Specifically, PLASMA does not support band matrices and
   does  not  solve  eigenvalue  and singular value problems. Also, PLASMA
   does   not   replace  ScaLAPACK  as  software  for  distributed  memory
   computers, since it only supports shared-memory machines.
     __________________________________________________________________

2. Where to Find More Information

   The main repository for PLASMA documentation is the distribution ./docs
   directory.  The  directory  contains  important  documents  such as the
   Users'  Guide  and  the  Reference Manual. PLASMA documentation is also
   available online on the PLASMA website:
   [4]http://icl.cs.utk.edu/plasma/.  For installation instructions please
   refer  to  the  [5]Installation  Guide. In addition, the [6]PLASMA User
   Forum  can be used to post general questions and comments as well as to
   report technical problems.
     __________________________________________________________________

3. Important Information about BLAS and LAPACK

  3.1. Optimized BLAS are Critical for Performance

   It  is absolutely critical for performance to use PLASMA in conjunction
   with   an   optimized   implementation  of  the  Basic  Linear  Algebra
   Subroutines  (BLAS)  library. Such implementations are usually provided
   by  the processor manufacturer and are usually available free of charge
   for non-profit use, such as academic research. Examples include:
     * The VecLib from Apple,
     * The AMD Core Math Library (ACML),
     * The Math Kernel Library (MKL) from Intel,
     * The Engineering and Scientific Software Library (ESSL) from IBM.

   Open-source alternatives also exist, such as:
     * [7]Goto BLAS,
     * [8]Automatically Tuned Linear Algebra Software (ATLAS).

   As  the  last resort, the FORTRAN implementation of BLAS from [9]Netlib
   can  be  used  (often  referred  to  as reference BLAS). However, since
   Netlib  BLAS  are  completely unoptimized, PLASMA with Netlib BLAS will
   deliver correct numerical results, but no performance whatsoever.

   For   comprehensive  installation  instructions  please  refer  to  the
   [10]Installation  Guide. However, if you decide to install manually and
   edit the installation scripts then there is one important issue to keep
   in  mind.  Modern  optimized  BLAS  are  not  stand-alone libraries but
   instead  are  bundled  with additional software, primarily LAPACK. This
   makes  it  necessary to use proper linking flags. Commonly these are -L
   (for path to library) and -l (for library name):
-L/usr/lib -lblas

   This will only pull in symbols from the BLAS library that are needed by
   PLASMA.  The  commonly made mistake is to just specify the full path to
   the library:
/usr/lib/libblas.a

   This  method  will  work  for  simple cases but it forces the linker to
   include the whole contents of the BLAS library rather than just pull in
   the  missing  symbols.  Aside  from  making the binary executable files
   larger,  this  method will easily cause name clashes as the same symbol
   name might be included multiple times.

  3.2. Multithreading within BLAS Must be Disabled

   Many  Basic  Linear  Algebra Subroutines (BLAS) implementations exploit
   parallelism   within  BLAS  through  multithreading.  PLASMA,  however,
   utilizes  BLAS  for  high  performance  implementations  of single-core
   operations  (often  referred to as kernels) and exploits parallelism at
   the  algorithmic level above the level of BLAS. For that reason, PLASMA
   must  not  be used in conjunction with a multithreaded BLAS, as this is
   likely  to  create  more  execution  threads  than  actual  cores.  The
   phenomenon,   known   as  oversubscribing  of  cores,  will  completely
   annihilate  PLASMA’s  performance  due  to  devastating impact on the
   operation of cache memories for dense linear algebra workloads.

   PLASMA  needs  to  be  linked  with  a  sequential  BLAS  library  or a
   multithreaded  BLAS  library with multithreading disabled. To learn how
   to  link  with sequential MKL, please consult the [11]Intel Math Kernel
   Library  Link Line Advisor; To learn how to link with sequential ATLAS,
   please  consult  the  [12]ATLAS  errata.  Alternatively,  PLASMA can be
   linked  with a multithreaded BLAS library with multithreading disabled.
   Typically,  disabling  of  multithreading  can  be  done by setting the
   appropriate environment variable from the command prompt, for instance:
export OMP_NUM_THREADS=1
export MKL_NUM_THREADS=1
export GOTO_NUM_THREADS=1

   Currently, this disables the simultaneous use of PLASMA for some of the
   user’s  functions  and  vendor BLAS for others. This cannot be easily
   remedied   without   standardization   of   interoperability  rules  of
   multithreaded   libraries.  One  consolation  is  that  PLASMA  already
   delivers fast parallel implementations of all Level 3 BLAS routines and
   there  is  virtually  no  benefit from parallelization of Level 1 and 2
   BLAS  routines  on  current  generation  of  multicore platforms due to
   memory contention.

  3.3. PLASMA Software Stack

   PLASMA  requires the following software packages to be installed in the
   system prior to PLASMA’s installation: BLAS, CBLAS, LAPACK and Netlib
   LAPACK C Wrapper. CBLAS and the components of LAPACK required by PLASMA
   are  commonly  bundled  with  BLAS.  If  this  is  not  the case Netlib
   implementation  of CBLAS and Netlib LAPACK can be used. The C interface
   to  LAPACK  has not been standardized yet and the LAPACK C Wrapper from
   Netlib has to be used for the time being.

   BLAS  is  the  set of Basic Linear Algebra Subprograms described in the
   previous  section.  An  unoptimized reference implementation of BLAS is
   available  from Netlib at [13]http://www.netlib.org/blas/. As mentioned
   before, it is critical for performance that optimized implementation of
   BLAS is used instead of Netlib BLAS.

   CBLAS  is  the  C  language  interface  to  BLAS  available  with  most
   implementations of BLAS. A reference implementation from Netlib is also
   available    at   [14]http://www.netlib.org/blas/blast-forum/cblas.tgz.
   Since  CBLAS  is  only  a set of wrappers to the actual BLAS, the CBLAS
   from Netlib can be used without any adverse effects on performance.

   LAPACK  is  a large package of linear algebra routines for a wide range
   of  problems.  PLASMA uses only a tiny portion of LAPACK, which is also
   commonly  bundled with BLAS distributions. If this is not the case, the
   complete  LAPACK  distribution  from  Netlib  can  be  used,  which  is
   available  at [15]http://www.netlib.org/lapack/. Although vendor LAPACK
   routines  can be more optimized than those from Netlib, there should be
   no  adverse  performance  effects  of using Netlib LAPACK, since PLASMA
   only relies on LAPACK for implementing some of its sequential kernels.

   The  user  can point PLASMA’s installer to all the components already
   installed  in  the  system.  For  all  the  missing  components  Netlib
   equivalents  will  be  installed.  The  installer can also be forced to
   disregard  any  software  already  installed  in the system and use the
   Netlib packages instead.

  3.4. Thou Shalt Not Mix Compilers

   For  a  given  processor,  the user can have different compilers at his
   disposal.  For instance, GNU, PGI and Intel compilers are available for
   Intel  processors.  Different  compilers  can  have  slightly different
   Application  Binary Interfaces (ABIs) and mixing compilers is generally
   a bad idea. User’s code and the PLASMA library should be compiled with
   the  same  compiler,  and so should be BLAS, CBLAS, LAPACK and LAPACK C
   Wrapper,  if a source distribution is used. If a binary distribution of
   the  BLAS  is  used,  the  correct  version  has  to be chosen (the one
   providing  the  right  ABI).  For  Intel processors, the [16]Intel Math
   Kernel Library Link Line Advisor can be used to assist with the choice.

  3.5. Linking FORTRAN code with C Code

   Currently  PLASMA  library  does not contain any FORTRAN code any more.
   FORTRAN is only used in PLASMA’s testing suite derived from the one of
   LAPACK (/testing/lin/). Because of that, neither a FORTRAN compiler nor
   the FORTRAN standard library has to be involved in compiling PLASMA and
   linking  it  with  C code. The following paragraph is preserved in this
   README only because it is a very useful piece of information for novice
   users who are forced to mix FORTRAN and C.

   If  FORTRAN code is mixed with C code, the FORTRAN standard library has
   to  be included. Sometimes it can be accomplished by simply putting the
   standard   FORTRAN   library  at  the  and  of  the  link  line,  e.g.,
   "-lgfortran"  when  using  GCC.  Alternatively, FORTRAN compiler can be
   used  for  linking. This will accomplish the same effect automatically.
   However,  the Intel IFORT compiler, when used for linking, assumes that
   the   main  program  is  in  FORTRAN  and  links  for_main.o  into  the
   application.  This  provides  the linker with two main() functions (one
   created  by  the  user  and one inserted by the build system), which is
   cause a linker error. To prevent this from happening, the "-nofor_main"
   link option has to be given.
     __________________________________________________________________

4. A Note on Running on NUMA Systems

   PLASMA  is a software package for shared memory systems, both Symmetric
   Multi-Processors  (SMP)  and  Non-Uniform Memory Access (NUMA) systems.
   PLASMA  does  not  detect  the type of the system and does not take any
   specific actions in that respect. PLASMA’s performance may be poor on
   NUMA  systems  if  matrices  are  not distributed among multiple memory
   nodes.  The  current  wisdom  is to use "numactl --interleave=all" when
   running an application that uses PLASMA.
     __________________________________________________________________

5. A Note on Running PLASMA and OpenMP

   PLASMA  currently binds the existing threads to specific cores in order
   to  optimize data locality during computation. As a consequence, if you
   add  your  first  OpenMP  section  after  the  call to PLASMA_Init, all
   threads  created  by the OpenMP section will be sons of the main thread
   binded on the core 0 and will thus be binded on the same core. To avoid
   this  problem,  it  is  recommended  to first have an OpenMP section to
   create  the  threads even if it does not do anything and then place the
   call  to  PLASMA_Init.  If  your  OpenMP  section  is after the call to
   PLASMA_Finalize,  it  should not be a problem, since version 2.4.1 will
   unbind threads after this call.
     __________________________________________________________________

6. License Information

   PLASMA  is  a  software  package  provided  by University of Tennessee,
   University  of California, Berkeley and University of Colorado, Denver.
   PLASMA’s  license  is  a  BSD-style  permissive free software license
   (properly  called  modified BSD). It allows proprietary commercial use,
   and for the software released under the license to be incorporated into
   proprietary  commercial  products.  Works  based on the material may be
   released  under  a  proprietary  license  as long as PLASMA’s license
   requirements  are maintained, as stated in the LICENSE file, located in
   the  main directory of the PLASMA distribution. In contrast to copyleft
   licenses, like the GNU General Public License, PLASMA’s license allows
   for  copies  and derivatives of the source code to be made available on
   terms more restrictive than those of the original license.
     __________________________________________________________________

7. Publications

   A  number  of  technical reports were written during the development of
   PLASMA  and  published as [17]LAPACK Working Notes by the University of
   Tennessee.  Almost  all  of  these  reports  later  appeared as journal
   articles.

   One  convenient way to make a reference to PLASMA is to cite [18]PLASMA
   Users'  Guide, Version 2.3, also available as a [19]LAPACK Working Note
   XXX and a [20]Technical Report UT-CS-XX-XXX. The following BibTeX entry
   can be used for LaTeX documents:

   @TECHREPORT{plasma_users_guide,
       author       = {{Dongarra}, J. and {Kurzak}, J. and {Langou}, J. and
                       {Langou}, J. and {Ltaief), H. and {Luszczek}, P. and
                       {YarKhan}, A. and {Alvaro}, W. and {Faverge}, M. and
                       {Haidar}, A. and {Hoffman}, J. and {Agullo}, E. and
                       {Buttari}, A. and {Hadri}, B.},
       title        = {{PLASMA} Users' Guide, Version 2.3},
       institution  = {Electrical Engineering and Computer Science Department,
                       Univesity of Tennessee},
       year         = {2010},
       type         = {technical report},
       number       = {UT-CS-XX-XXX},
       address      = {Knoxville, Tennessee 37996},
       month        = {September},
       note         = {},
       key          = {},
   }
     __________________________________________________________________

8. Funding

   The   PLASMA  project  is  funded  in  part  by  the  National  Science
   Foundation, Department of Energy, Microsoft, and the MathWorks.
     __________________________________________________________________

   Last updated 2011-09-14 13:39:40 EDT

Références

   1. http://icl.cs.utk.edu/plasma/
   2. http://www.netlib.org/lapack/
   3. http://www.netlib.org/scalapack/
   4. http://icl.cs.utk.edu/plasma/
   5. http://icl.cs.utk.edu/projectsfiles/plasma/html/InstallationGuide.html
   6. http://icl.cs.utk.edu/plasma/forum/
   7. http://web.tacc.utexas.edu/~kgoto/
   8. http://math-atlas.sourceforge.net/
   9. http://www.netlib.org/blas/
  10. http://icl.cs.utk.edu/projectsfiles/plasma/html/InstallationGuide.html
  11. http://software.intel.com/en-us/articles/intel-mkl-link-line-advisor/
  12. http://math-atlas.sourceforge.net/errata.html#LINK
  13. http://www.netlib.org/blas/
  14. http://www.netlib.org/blas/blast-forum/cblas.tgz
  15. http://www.netlib.org/lapack/
  16. http://software.intel.com/en-us/articles/intel-mkl-link-line-advisor/
  17. http://www.netlib.org/lapack/lawns/downloads/
  18. http://icl.cs.utk.edu/projectsfiles/plasma/pdf/users_guide.pdf
  19. http://www.netlib.org/lapack/lawnspdf/lawnXXX.pdf
  20. http://www.cs.utk.edu/~library/TechReports/2010/ut-cs-XX-XXX.pdf


ReleaseNotes
                              PLASMA RELEASE NOTES
     __________________________________________________________________

Summary of Current Features

     * Solution  of  dense  systems  of  linear equations and least square
       problems  in  real  space and complex space, using single precision
       and   double   precision,   via   the   Cholesky,  LU,  QR  and  LQ
       factorizations
     * Solution  of  standard  and  generalized dense symmetric Eigenvalue
       Problem in real space and complex space, using single precision and
       double  precision,  via two-stage tridiagonal reduction followed by
       QR iteration (eigenvalues only, no eigenvectors yet)
     * Dense Singular Value Decomposition in real space and complex space,
       using   single   precision  and  double  precision,  via  two-stage
       bidiagonal  reduction  followed  by  QR  iteration (singular values
       only, no singular vectors yet)
     * Solution  of  dense  linear  systems of equations in real space and
       complex  space  using  the  mixed-precision  algorithm based on the
       Cholesky, LU, QR and LQ factorizations
     * Tree-based  QR  and  LQ  factorizations and Q matrix generation and
       application (“tall and skinny”)
     * Tree-based bidiagonal reduction (“tall and skinny”)
     * Explicit   matrix   inversion   based   on  Cholesky  factorization
       (symmetric positive definite)
     * Parallel   and   cache-efficient   in-place   layout   translations
       (Gustavson et at.)
     * Complete  set  of Level 3 BLAS routines for matrices stored in tile
       layout
     * Simple  LAPACK-like interface for greater productivity and advanced
       (tile) interface for full control and maximum performance; Routines
       for  conversion  between  LAPACK  matrix layout and PLASMA’s tile
       layout
     * Dynamic  scheduler  QUARK  (QUeuing  And  Runtime  for Kernels) and
       dynamically   scheduled  versions  of  all  computational  routines
       (alongside statically scheduled ones)
     * Asynchronous interface for launching dynamically scheduled routines
       in  a  non-blocking  mode.  Sequence  and  request  constructs  for
       controlling progress and checking errors
     * Automatic  handling  of  workspace  allocation;  A set of auxiliary
       functions to assist the user with workspace allocation
     * A simple set of "sanity" tests for all numerical routines including
       Level 3 BLAS routines for matrices in tile layout
     * An  advanced  testing suite for exhaustive numerical testing of all
       the  routines  in all precisions (based on the testing suite of the
       LAPACK library)
     * Basic timing suite for most of the routines in all precisions
     * Thread safety
     * Support for Make and CMake build systems
     * LAPACK-style comments in the source code using the Doxygen system
     * Native  support  for  Microsoft  Windows using WinThreads through a
       thin OS interaction layer
     * Installer capable of downloading from Netlib and installing missing
       components of PLASMA’s software stack (BLAS, CBLAS, LAPACK, LAPACKE
       C API)
     * Extensive documentation including Installation Guide, Users' Guide,
       Reference  Manual  and  an  HTML  code  browser, a guide on running
       PLASMA  with  the  TAU  package,  Contributors' Guide, a README and
       Release Notes.
     * A comprehensive set of usage examples
     __________________________________________________________________

New Features by Release

  2.4.2, September 14th, 2011

     * New  version  of quark removing active waiting and allowing user to
       bind tasks to set of cores.
     * Installer:  Fix  compatibility  issues between plasma-installer and
       PGI compiler reported on Kraken by A. Bouteiller.
     * Fix one memory leak with Hwloc.
     * Introduce  a  new  kernel  for  the  recursive LU operation on tile
       layout which reduces cache misses.
     * Fix  several  bugs and introduce new features thanks to people from
       Fujitsu and NAG :
          + The  new  LU factorization with partial pivoting introduced in
            release 2.4 is now working on rectangular matrices.
          + Add missing functions to Fortran 77 interface.
          + Add  a  new  Fortran  90  interface  to  all  LAPACK  and Tile
            interface. Asynchronous interface and mixed precision routines
            are not available yet.
          + Fix arguments order in header files to fit implementation.

  2.4.1, July 8th, 2011

     * Fix   bug   with   Fujitsu   compiler   reported   on   the   forum
       ([1]http://icl.cs.utk.edu/plasma/forum/viewtopic.php?f=2&t=108)
     * Unbind  threads  in  PLASMA_Finalize to avoid problem of binding in
       OpenMP  section  following  PLASMA calls (still possible on Mac and
       AIX  without hwloc). A better fix is to create the OpenMP thread in
       the  user  code before any call to PLASMA thanks to a fake parallel
       section.

  2.4.0, June 6th, 2011

     * Tree-based  QR  and  LQ factorizations: routines for application of
       the   Q  matrix  support  all  combinations  of  input  parameters:
       Left/Right, NoTrans/Trans/ConjTrans
     * Symmetric    Engenvalue    Problem   using:   tile   reduction   to
       band-tridiagonal  form, reduction to "standard" tridiagonal form by
       bulge   chasing,   finding   eigenvalues  using  the  QR  algorithm
       (eigenvectors currently not supported)
     * Singular    Value    Decomposition   using:   tile   reduction   to
       band-bidiagonal form, reduction to “standard” bidiagonal form by
       bulge  chasing,  finding  singular  values  using  the QR algorithm
       (singular vectors currently no supported)
     * Gaussian  Elimination  with  partial  pivoting  (as  opposed to the
       incremental  pivoting  in  the  tile LU factorization) and parallel

quark/quark.h
/**
 *
 * @file quark.h
 *
 * Dynamic scheduler functions
 *
 * PLASMA is a software package provided by Univ. of Tennessee,
 * Univ. of California Berkeley and Univ. of Colorado Denver
 *
 * @version 2.4.2
 * @author Asim YarKhan
 * @date 2010-11-15
 *
 **/

#ifndef QUARK_H
#define QUARK_H

#include <limits.h>
#include <stdio.h>

#if defined( _WIN32 )
  /* This must be included before INPUT is defined below, otherwise we
     have a name clash/problem  */
  #include <windows.h>
  #include <limits.h>
#else
  #include <inttypes.h>
#endif

#include "quark_unpack_args.h"

#if defined(c_plusplus) || defined(__cplusplus)
extern "C" {
#endif

#ifdef DBGQUARK
/* #define DBGPRINTF(str, ...) { fprintf(stderr, "%s:%d: [%s] " str, __FILE__, __LINE__, __FUNCTION__, __VA_ARGS__); } */
#define DBGPRINTF(...) { fprintf(stderr, __VA_ARGS__); }
#else
#define DBGPRINTF(...) if (0) {};
#endif

#define QUARK_SUCCESS 0
#define QUARK_ERR -1
#define QUARK_ERR_UNEXPECTED -1
#define QUARK_ERR_NOT_SUPPORTED -2

/* A bitmask of 8 bits to to hold region markers  */
#define QUARK_REGION_BITMASK 0x0000FF
#define QUARK_REGION_ALL 0x0FF
typedef enum { QUARK_REGION_0=1<<0, QUARK_REGION_1=1<<1, QUARK_REGION_2=1<<2, QUARK_REGION_3=1<<3,
               QUARK_REGION_4=1<<4, QUARK_REGION_5=1<<5, QUARK_REGION_6=1<<6, QUARK_REGION_7=1<<7 } quark_data_region_t;
typedef enum { QUARK_REGION_L=QUARK_REGION_0|QUARK_REGION_1|QUARK_REGION_2,
               QUARK_REGION_D=QUARK_REGION_3|QUARK_REGION_4,
               QUARK_REGION_U=QUARK_REGION_5|QUARK_REGION_6|QUARK_REGION_7 } quark_ldu_region_t;

/* Data items can be: */
/*  INPUT, OUTPUT, INOUT:  these data items create dependencies */
/*  VALUE:  these data items get copied over */
/*  NODEP:  these data items get copied over, and are not used for dependencies */
/*  SCRATCH:  these data items can be allocated (and deallocted) by the scheduler when tasks execute  */
#define QUARK_DIRECTION_BITMASK 0x000F00
typedef enum { QINPUT=0x100, OUTPUT=0x200, INOUT=0x300, VALUE=0x400, NODEP=0x500, SCRATCH=0x600} quark_direction_t;
#define INPUT 0x100

#define QUARK_VALUE_FLAGS_BITMASK 0xFFF000

/* Data locality flag; ie keep data on the same core if possible */
#define LOCALITY ( 1 << 12 )
#define NOLOCALITY 0x00

/* A data address with a sequence of ACCUMULATOR dependencies will allow the related tasks to be reordered  */
#define ACCUMULATOR ( 1 << 13 )
#define NOACCUMULATOR 0x00

/* A data address with a sequence of GATHERV dependencies will allow the related tasks to be run in parallel  */
#define GATHERV ( 1 << 14 )
#define NOGATHERV 0x00

/* The following are task level flags, that can be either provided as additional arguments to the task, or via SET functions */
/* The task label; should be provided as a null terminated string */
#define TASK_LABEL ( 1 << 15 )
#define TASKLABEL TASK_LABEL    /* depreciated label */
/* The task color; should be provided as a null terminated string */
#define TASK_COLOR ( 1 << 16 )
#define TASKCOLOR TASK_COLOR    /* depreciated label */
/* The priority of the task, provided as an integer */
#define TASK_PRIORITY ( 1 << 17 )
/* Lock the task to a specific thread number (0 ... NTHREADS-1), provided as an integer */
#define TASK_LOCK_TO_THREAD ( 1 << 18 )
/* The sequence pointer to be associated with the task, provided as a pointer */
#define TASK_SEQUENCE ( 1 << 19 )
/* An integere for the number of threads require */
#define TASK_THREAD_COUNT ( 1 << 20 )
/* The thread that runs this task should have manual scheduling enabled (1) or disabled (0) */
#define THREAD_SET_TO_MANUAL_SCHEDULING ( 1 << 21 )
/* Lock the task to a thead mask (0 ... NTHREADS-1) bits long, provided as a character array (byte array) */
#define TASK_LOCK_TO_THREAD_MASK ( 1 << 22 )

/* The range for priority values */
#define QUARK_TASK_MIN_PRIORITY 0
#define QUARK_TASK_MAX_PRIORITY INT_MAX

/* Definition of structure holding scheduler information */
typedef struct quark_s Quark;

/* Structure holding task information */
typedef struct quark_task_s Quark_Task;

/* Create a type for setting task flags */
struct quark_task_flags_s {
    int task_priority;
    int task_lock_to_thread;
    char *task_color;
    char *task_label;
    void *task_sequence;
    int task_thread_count;
    int thread_set_to_manual_scheduling;
    unsigned char *task_lock_to_thread_mask;
};

typedef struct quark_task_flags_s Quark_Task_Flags;
/* Static initializer for Quark_Task_Flags_t */
#define Quark_Task_Flags_Initializer { (int)0, (int)-1, (char *)NULL, (char *)NULL, (void *)NULL, (int)1, (int)-1, (unsigned char *)NULL }

/* Setup scheduler data structures, assumes threads are managed seperately */
Quark *QUARK_Setup(int num_threads);

/* Setup scheduler data structures, spawn worker threads, start the workers working  */
Quark *QUARK_New(int num_threads);

/* Add a task, called by the master process (thread_rank 0)  */
unsigned long long QUARK_Insert_Task(Quark * quark, void (*function) (Quark *), Quark_Task_Flags *task_flags, ...);

/* Main work loop, called externally by everyone but the master
 * (master manages this internally to the insert_task and waitall
 * routines). Each worker thread can call work_main_loop( quark,
 * thread_rank), where thread rank is 1...NUMTHREADS ) */
void QUARK_Worker_Loop(Quark *quark, int thread_rank);

/* Finish work and return.  Workers do not exit */
void QUARK_Barrier(Quark * quark);

/* Just wait for current tasks to complete, the scheduler and
 * strutures remain as is... should allow for repeated use of the
 * scheduler.  The workers return from their loops.*/
void QUARK_Waitall(Quark * quark);

/* Delete scheduler, shutdown threads, finish everything, free structures */
void QUARK_Delete(Quark * quark);

/* Free scheduling data structures */
void QUARK_Free(Quark * quark);

/* Cancel a specific task */
int QUARK_Cancel_Task(Quark *quark, unsigned long long taskid);

/* Returns a pointer to the list of arguments, used when unpacking the
   arguments; Returna a pointer to icl_list_t, so icl_list.h will need
   bo included if you use this function */
void *QUARK_Args_List(Quark *quark);

/* Returns the rank of a thread in a parallel task */
int QUARK_Get_RankInTask(Quark *quark);

/* Return a pointer to an argument.  The variable last_arg should be
   NULL on the first call, then each subsequent call will used
   last_arg to get the the next argument. */
void *QUARK_Args_Pop( void *args_list, void **last_arg);

/* Utility function returning rank of the current thread */
int QUARK_Thread_Rank(Quark *quark);

/* Packed task interface */
/* Create a task data structure to hold arguments */
Quark_Task *QUARK_Task_Init(Quark * quark, void (*function) (Quark *), Quark_Task_Flags *task_flags );

/* Add (or pack) the arguments into a task data structure (make sure of the correct order) */
void QUARK_Task_Pack_Arg( Quark *quark, Quark_Task *task, int arg_size, void *arg_ptr, int arg_flags );

/* Insert the packed task data strucure into the scheduler for execution */
unsigned long long QUARK_Insert_Task_Packed(Quark * quark, Quark_Task *task );

/* Unsupported function for debugging purposes; execute task AT ONCE */
unsigned long long QUARK_Execute_Task(Quark * quark, void (*function) (Quark *), Quark_Task_Flags *task_flags, ...);

/* Get the label (if any) associated with the current task; used for printing and debugging  */
char *QUARK_Get_Task_Label(Quark *quark);

/* Method for setting task flags */
Quark_Task_Flags *QUARK_Task_Flag_Set( Quark_Task_Flags *flags, int flag, intptr_t val );

/* Type for task sequences */
typedef struct Quark_sequence_s Quark_Sequence;

/* Create a seqeuence structure, to hold sequences of tasks */
Quark_Sequence *QUARK_Sequence_Create( Quark *quark );

/* Called by worker, cancel any pending tasks, and mark sequence so that it does not accept any more tasks */
int QUARK_Sequence_Cancel( Quark *quark, Quark_Sequence *sequence );

/* Destroy a sequence structure, cancelling any pending tasks */
Quark_Sequence *QUARK_Sequence_Destroy( Quark *quark, Quark_Sequence *sequence );

/* Wait for a sequence of tasks to complete */
int QUARK_Sequence_Wait( Quark *quark, Quark_Sequence *sequence );

/* Get the sequence information associated the current task/worker, this was provided when the tasks was created */
Quark_Sequence *QUARK_Get_Sequence(Quark *quark);

/* Get the priority associated the current task/worker */
int QUARK_Get_Priority(Quark *quark);

/* Get information associated the current task and worker thread;
 * Callable from within a task, since it works on the currently
 * executing task */
intptr_t QUARK_Task_Flag_Get( Quark *quark, int flag );

/* Enable and disable DAG generation via API.  Only makes sense after
 * a sync, such as QUARK_Barrier. */
void QUARK_DOT_DAG_Enable( Quark *quark, int boolean_value );

/* Get the number_th bit in a bitset (unsigned char *); useful for QUARK_LOCK_TO_THREAD_MASK flag */
static inline int QUARK_Bit_Get(unsigned char *set, int number)
{
    set += number / 8;
    return (*set & (1 << (7-(number % 8)))) != 0; /* 0 or 1       */
}

/* Set the number_th bit in a bitset (unsigned char *) to value (0 or 1); useful for QUARK_LOCK_TO_THREAD_MASK flag */
static inline void QUARK_Bit_Set(unsigned char *set, int number, int value)
{
    set += number / 8;
    if (value)
        *set |= 1 << (7-(number % 8)); /* set bit      */
    else    *set &= ~(1 << (7-(number % 8))); /* clear bit    */

}

#if defined(c_plusplus) || defined(__cplusplus)
}
#endif

#endif                          /* QUARK.H */
